spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoClientFactorySupport.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.boot.autoconfigure.mongo;
+
+import java.util.Collections;
+import java.util.List;
+
+import com.mongodb.ConnectionString;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.MongoClientSettings.Builder;
+import com.mongodb.MongoCredential;
+import com.mongodb.MongoDriverInformation;
+import com.mongodb.ServerAddress;
+
+import org.springframework.core.env.Environment;
+import org.springframework.util.Assert;
+
+/**
+ * Base class for common setup bits (aka {@link MongoClientSettings}) required for
+ * instantiating a MongoClient.
+ *
+ * @author Christoph Strobl
+ * @since 2.3.0
+ */
+public abstract class MongoClientFactorySupport<T> {
+
+	private final MongoProperties properties;
+
+	private final Environment environment;
+
+	private final List<MongoClientSettingsBuilderCustomizer> builderCustomizers;
+
+	public MongoClientFactorySupport(MongoProperties properties, Environment environment,
+			List<MongoClientSettingsBuilderCustomizer> builderCustomizers) {
+		this.properties = properties;
+		this.environment = environment;
+		this.builderCustomizers = (builderCustomizers != null) ? builderCustomizers : Collections.emptyList();
+	}
+
+	public T createMongoClient(MongoClientSettings settings) {
+		MongoClientSettings targetSettings = computeClientSettings(settings);
+		return (getEmbeddedPort() != null) ? createEmbeddedMongoClient(targetSettings)
+				: createNetworkMongoClient(targetSettings);
+	}
+
+	private MongoClientSettings computeClientSettings(MongoClientSettings settings) {
+
+		Builder settingsBuilder = (settings != null) ? MongoClientSettings.builder(settings)
+				: MongoClientSettings.builder();
+		applyHostAndPort(settingsBuilder);
+		applyCredentials(settingsBuilder);
+
+		customize(settingsBuilder);
+		return settingsBuilder.build();
+	}
+
+	private void applyHostAndPort(MongoClientSettings.Builder settings) {
+
+		if (isEmbedded()) {
+			settings.applyConnectionString(new ConnectionString("mongodb://localhost:" + getEmbeddedPort()));
+			return;
+		}
+		if (!this.properties.determineUri().equals(MongoProperties.DEFAULT_URI)) {
+			if (hasCustomAddress()) {
+				Assert.state(this.properties.getUri() == null,
```

Comment:
This is breaking an existing test and I fail to see how that's related to the migration to the new driver. Can you please clarify?

Reply:
It seems to be a mismatch that existed between the reactive and imperative implementation. If you remove that `if` statement `ReactiveMongoClientFactoryTests.uriCannotBeSetWithHostPort` will fail which contradicts `MongoPropertiesTests.uriOverridesHostAndPort`.